# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: js/src/jit/JitFrames.cpp
# Commit: 6ccaefced3df
# Full Hash: 6ccaefced3df63b297057b31a856527153801a71
# Author: Jan de Mooij <jdemooij@mozilla.com>
# Date: 2020-01-13 15:45:24
# Description:
#   Bug 1605641 - Mark 'done' stack value for destructuring assignment as having implicit uses. r=tcampbell
#   
#   We need to make sure this value isn't optimized out because the exception
#   handler may need it.
#   
# ==============================================================================

diff -r 96753184f3c4 -r 6ccaefced3df js/src/jit/JitFrames.cpp
--- a/js/src/jit/JitFrames.cpp	Thu Jan 02 12:07:51 2020 +0000
+++ b/js/src/jit/JitFrames.cpp	Mon Jan 13 09:23:17 2020 +0000
@@ -114,6 +114,7 @@
 
   if (isDestructuring) {
     RootedValue doneValue(cx, si.read());
+    MOZ_RELEASE_ASSERT(!doneValue.isMagic());
     bool done = ToBoolean(doneValue);
     // Do not call IteratorClose if the destructuring iterator is already
     // done.
@@ -399,7 +400,10 @@
         uint8_t* stackPointer;
         BaselineFrameAndStackPointersFromTryNote(tn, frame, &framePointer,
                                                  &stackPointer);
+        // Note: if this ever changes, also update the JSTRY_DESTRUCTURING code
+        // in IonBuilder.cpp!
         RootedValue doneValue(cx, *(reinterpret_cast<Value*>(stackPointer)));
+        MOZ_RELEASE_ASSERT(!doneValue.isMagic());
         bool done = ToBoolean(doneValue);
         if (!done) {
           Value iterValue(*(reinterpret_cast<Value*>(stackPointer) + 1));